### PR TITLE
History callbacks

### PR DESCRIPTION
**Pull request by olgavrou**
#199 

- added on request callback to history (trying to be consistent with the way functions are registered in ccf so far, please let me know if it should be done in a different way)
- added the client proxy in pbft.h
- pbft_config.h added as helper to configure the client proxy and in wrapping the ccf requests (more work to come on the request wrapping)

Please keep in mind that this is an intermediate step towards integration so hopefully things will start to look better soon.

Some tests are currently failing due to a combination of forwarding and pbft not being in sync with raft. In next PR I want to have pbft_config.h include an option to run pbft standalone using ccf for node2node communication  

TODOs:
- Pass the RequestID as caller_rid (probably need to add hash for tuples in ds/hash.h)
- default option to have pbft run standalone (i.e. generating it's own requests) so that it doesn't rely on raft. Useful for debugging pbft inside the enclave when needed.

### CODE REVIEW

File: src/pbft/pbft_config.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      {
+        LOG_FATAL << "total requests executed: " << total_requests_executed
+                  << " not equal to exec command counter: " << counter << "\n";
+        throw std::logic_error(
+          "Total requests executed not equal to exec command counter");
+      }
+
+      if (total_requests_executed % 100 == 0)
+      {
+        LOG_INFO << "total requests executed " << total_requests_executed
```
Comment (eddyashton) [inline src/pbft/pbft_config.h]: Can you replace these macros with the `LOG_INFO_FMT` variants? I think they're more readable, and we should be consistent.

File: src/kv/kvtypes.h [hunk 2]
```diff
@@ -61,21 +61,35 @@ namespace kv
   class TxHistory
   {
   public:
-    using RequestID = std::tuple<size_t /* Caller ID */, size_t /* Client Session ID */, kv::Version>;
+    using RequestID = std::tuple<
```
Comment (eddyashton) [inline src/kv/kvtypes.h]: (Not really part of this PR but) This should probably be a struct? Then we only need a std::hash for this type, not for all tuples.

File: src/pbft/pbft_config.h (hunk 1 above)
Comment (eddyashton) [reply src/pbft/pbft_config.h]: Actually this branch may have been taken before the `_FMT` variants were added, in which case ignore this.

File: src/pbft/pbft_config.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+// Licensed under the Apache 2.0 License.
+#include "libbyz/libbyz.h"
+#include "libbyz/pbft_assert.h"
+#include "pbft/pbft_deps.h"
+
+namespace pbft
+{
+  static char* service_mem = 0;
+
+  class PbftConfig
```
Comment (jumaffre) [inline src/pbft/pbft_config.h]: Minor: `AbstractPbftConfig` to keep it consistent with other interface classes in CCF?

File: src/node/history.h [hunk 4]
```diff
@@ -316,21 +334,36 @@ namespace ccf
...
 
-    void add_request(kv::TxHistory::RequestID id, const std::vector<uint8_t>& request) override
+    void add_request(
+      kv::TxHistory::RequestID id, const std::vector<uint8_t>& request) override
     {
       LOG_DEBUG << fmt::format("HISTORY: add_request {0}", id) << std::endl;
       requests[id] = request;
+#ifdef PBFT
+      auto callback = callbacks.find(pbft::Callbacks::ON_REQUEST);
+      if (callback != callbacks.end())
```
Comment (jumaffre) [inline src/node/history.h]: Should this throw an error or report a failure if this condition isn't true?

File: src/kv/kvtypes.h [hunk 5]
```diff
@@ -61,21 +61,35 @@ namespace kv
...
+      size_t /* Caller ID */,
+      size_t /* Client Session ID */,
+      kv::Version>;
```
Comment (jumaffre) [inline src/kv/kvtypes.h]: There is an inconsistency here between the third element of the tuple: here defined as a `kv::Version` but in `frontend.h`, we initialise it to the JSON-RPC sequence number (`reqid = {caller_id.value(), ctx.client_session_id, jsonrpc_id};`)

File: src/pbft/pbft_config.h [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    size_t message_size() override
+    {
+      return sizeof(ccf_req);
+    }
+
+    void fill_request(
+      uint8_t* buffer,
+      const size_t& total_req_size,
+      const std::vector<uint8_t>& data,
+      const size_t& version) override
```
Comment (jumaffre) [inline src/pbft/pbft_config.h]: `version` can be passed by value here.

File: src/pbft/pbft.h [hunk 7]
```diff
@@ -129,25 +135,54 @@ namespace pbft
...
+        cp->recv_reply(m);
+      };
+
+      message_receiver_base->register_reply_handler(cb, client_proxy.get());
+
+      callbacks[pbft::Callbacks::ON_REQUEST] =
+        [&](kv::TxHistory::CallbackArgs args) {
+          auto caller = std::get<0>(args.id);
+          auto session = std::get<1>(args.id);
+          auto version = std::get<2>(args.id);
```
Comment (jumaffre) [inline src/pbft/pbft.h]: As per my previous comment, this cannot be `version` at this point since this is called before the execution of the request.

File: src/pbft/pbft.h [hunk 8]
```diff
@@ -129,25 +135,54 @@ namespace pbft
...
-        exec_command,
+        pbft_config->get_exec_command(),
         0,
         0,
         pbft_network.get(),
         &message_receiver_base);
+
+      service_mem = mem + used_bytes;
+
+      static std::unique_ptr<ClientProxy<uint64_t, void>> client_proxy;
```
Comment (jumaffre) [inline src/pbft/pbft.h]: `auto client_proxy = std::make_unique<ClientProxy<uint64_t, void>>(*message_receiver_base);` ?

Also, we could make use of `using` to make it clearer what `uint64_t` is.

File: src/pbft/pbft_config.h [hunk 9]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "libbyz/libbyz.h"
+#include "libbyz/pbft_assert.h"
+#include "pbft/pbft_deps.h"
+
+namespace pbft
+{
+  static char* service_mem = 0;
```
Comment (jumaffre) [inline src/pbft/pbft_config.h]: This global variable should be made local to the `PbftConfigCCF` class.

File: src/pbft/pbft.h [hunk 10]
```diff
@@ -129,25 +135,54 @@ namespace pbft
...
+      LOG_INFO << "Setting up client proxy " << std::endl;
+      client_proxy.reset(
+        new ClientProxy<uint64_t, void>(*message_receiver_base));
+
+      auto cb = [](Reply* m, void* ctx) {
+        auto cp = (ClientProxy<uint64_t, void>*)ctx;
```
Comment (jumaffre) [inline src/pbft/pbft.h]: `static_cast`?

File: src/pbft/pbft_config.h [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        memcpy(&data[j], &request->get_counter_array()[j], sizeof(uint8_t));
+      }
+
+      Byz_modify(outb->contents, 8);
+      bzero(outb->contents, 8);
+      outb->size = 8;
+      return 0;
+    };
+
+    ExecCommand get_exec_command() override
```
Comment (jumaffre) [inline src/pbft/pbft_config.h]: Should this be under `public:`?

File: src/pbft/pbft_config.h [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+  private:
+    ExecCommand exec_command = [](
+                                 Byz_req* inb,
+                                 Byz_rep* outb,
+                                 _Byz_buffer* non_det,
+                                 int client,
+                                 bool ro,
+                                 Seqno total_requests_executed) {
+      Long& counter = *(Long*)service_mem;
```
Comment (jumaffre) [inline src/pbft/pbft_config.h]: Could we write a `TODO` describing what `exec_command` is currently doing (inspired by the pbft counter example and nothing to do with CCF for now) and explain that this will change soon?

File: src/pbft/pbft_config.h (hunk 3 above)
Comment (achamayou) [reply src/pbft/pbft_config.h]: AbstractPBFTConfig?

File: src/kv/kvtypes.h (hunk 5 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Yes, this is definitely wrong.

File: src/crypto/hash.h [hunk 13]
```diff
@@ -59,10 +58,7 @@ namespace fmt
     template <typename FormatContext>
     auto format(const crypto::Sha256Hash& p, FormatContext& ctx)
     {
-      return format_to(
-        ctx.out(),
-        "<sha256 {:02x}>",
-        fmt::join(p.h, ""));
+      return format_to(ctx.out(), "<sha256 {:02x}>", fmt::join(p.h, ""));
```
Comment (achamayou) [inline src/crypto/hash.h]: Slightly surprised by this reformatting, are you using the right clang format?

File: src/kv/kvtypes.h (hunk 2 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: We don't hash it at the moment, and this was tweaked somewhat, but now that it's agreed on, yes!

File: src/node/history.h [hunk 14]
```diff
@@ -235,6 +248,11 @@ namespace ccf
       nodes(nodes_)
     {}
 
+    void register_callback(std::string name, CallbackHandler func) override
+    {
+      callbacks[name] = func;
```
Comment (achamayou) [inline src/node/history.h]: Do we need a map of those? It seems to me that we're always going to have the same three?

File: src/pbft/pbft_config.h [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+      return sizeof(ccf_req);
+    }
+
+    void fill_request(
+      uint8_t* buffer,
+      const size_t& total_req_size,
+      const std::vector<uint8_t>& data,
+      const size_t& version) override
+    {
+      auto request = new (buffer) ccf_req;
```
Comment (achamayou) [inline src/pbft/pbft_config.h]: Where is request freed?

File: src/pbft/pbft_config.h [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+      if (total_requests_executed % 100 == 0)
+      {
+        LOG_INFO << "total requests executed " << total_requests_executed
+                 << "\n";
+      }
+
+      LOG_INFO << "request inb size: " << inb->size << std::endl;
+
+      auto request = new (inb->contents) ccf_req;
```
Comment (achamayou) [inline src/pbft/pbft_config.h]: Where is request freed?

File: src/pbft/pbft_config.h (hunk 1 above)
Comment (olgavrou) [reply src/pbft/pbft_config.h]: We have pulled in master so I can update these, thanks!

File: src/kv/kvtypes.h (hunk 2 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: I can add this functionality in the next PR where I want to pass the request id into client proxy anyway

File: src/kv/kvtypes.h (hunk 5 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: Yes thank you, I will update this

File: src/crypto/hash.h (hunk 13 above)
Comment (olgavrou) [reply src/crypto/hash.h]: Yes, I think I ran the check-format.sh file with the formatting option on, but I can always check again

File: src/pbft/pbft_config.h (hunk 15 above)
Comment (olgavrou) [reply src/pbft/pbft_config.h]: It is placement new

File: src/pbft/pbft_config.h (hunk 16 above)
Comment (olgavrou) [reply src/pbft/pbft_config.h]: this is placement new, there is no need to free it here

File: src/pbft/pbft_config.h (hunk 11 above)
Comment (olgavrou) [reply src/pbft/pbft_config.h]: yes

File: src/kv/kvtypes.h (hunk 2 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Yes, that's fine
